lang/cs/Org.Apache.REEF.Network/Elastic/Topology/Physical/Default/OperatorTopologyWithDefaultCommunication.cs

```diff
@@ -0,0 +1,261 @@
+        /// </summary>
+        /// <param name="cancellationSource">The signal to cancel the operation</param>
+        public virtual void WaitForTaskRegistration(CancellationTokenSource cancellationSource)
+        {
+            try
+            {
+                _commLayer.WaitForTaskRegistration(_children.Values.ToList(), cancellationSource);
+            }
+            catch (Exception e)
+            {
+                throw new OperationCanceledException("Failed to find parent/children nodes in operator topology for node: " + TaskId, e);
+            }
+
+            _initialized = true;
+
+            // Some message may have been received while we were setting up the topology. Send them.
+            Send(cancellationSource);
+        }
+
+        /// <summary>
+        /// Block and wait untill a message is received.
+        /// </summary>
+        /// <param name="cancellationSource">The signal that the operation is cacelled</param>
+        /// <returns></returns>
+        public virtual ElasticGroupCommunicationMessage Receive(CancellationTokenSource cancellationSource)
+        {
+            ElasticGroupCommunicationMessage message;
+            int retry = 1;
+
+            while (!_messageQueue.TryTake(out message, _timeout, cancellationSource.Token))
+            {
+                if (cancellationSource.IsCancellationRequested)
+                {
+                    throw new OperationCanceledException("Received cancellation request: stop receiving.");
+                }
+
+                if (retry++ > _retry)
+                {
+                    throw new Exception($"Failed to receive message after {_retry} try.");
+                }
+            }
+
+            return message;
+        }
+
+        /// <summary>
+        /// Send the input message. This method is asynchornous.
+        /// </summary>
+        /// <param name="message">The message to communicate</param>
+        /// <param name="cancellationSource">The signal for cancelling the operation</param>
+        public virtual void Send(ElasticGroupCommunicationMessage message, CancellationTokenSource cancellationSource)
+        {
+            _sendQueue.Enqueue(message);
+
+            if (_initialized)
+            {
+                Send(cancellationSource);
+            }
+        }
+
+        /// <summary>
+        /// Handler for incoming messages from other topology nodes.
+        /// </summary>
+        /// <param name="message">The message that need to be devlivered to the operator</param>
+        public virtual void OnNext(NsMessage<ElasticGroupCommunicationMessage> message)
+        {
+            if (_messageQueue.IsAddingCompleted)
+            {
+                if (_messageQueue.Count > 0)
+                {
+                    throw new IllegalStateException("Trying to add messages to a closed non-empty queue.");
+                }
+
+                _messageQueue = new BlockingCollection<ElasticGroupCommunicationMessage>();
```

Comment:
Yea this is basically because I don't know how to deal with OnComplete or OnError from Wake. Ok let's remove the new, make `_messageQueue` readonly and simply let it throw if the queue is completed. Fixed.